### PR TITLE
Expose memory_reservation_locked_to_max via cloud_properties

### PR DESCRIPTION
**Pull request by rkoster**
# Description
By setting this property all guest memory will be reserved. If a VM has more memory reserved than is available in a given resource pool it will fail to power on.

For multi-cluster configurations where resource pool memory reservations have been set some logic has been added which will try the next available cluster in case of a VM power-on failure.

## Impacted Areas in Application
List general components of the application that this PR will affect:

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?
Integration tests have been successfully run against one of the pooled ci environments.

### CODE REVIEW

File: src/vsphere_cpi/lib/cloud/vsphere/vm_type.rb [hunk 1]
```diff
@@ -15,7 +15,7 @@ def initialize(datacenter, cloud_properties, pbm)
     end
 
     %w[cpu ram disk cpu_hot_add_enabled memory_hot_add_enabled nsx vmx_options
-       nsxt nested_hardware_virtualization upgrade_hw_version datacenters vm_group disable_drs storage_policy tags].each do |name|
+       nsxt memory_reservation_locked_to_max nested_hardware_virtualization upgrade_hw_version datacenters vm_group disable_drs storage_policy tags].each do |name|
```
Comment (cunnie) [inline src/vsphere_cpi/lib/cloud/vsphere/vm_type.rb:18]: Looks good.
I _believe_ that by setting this to true & attempting to power on the VM & if there's not enough memory (`cluster_free_memory` - `overall_usage`) it won't power on.

We should, at some point, update the bosh.io docs to reflect this new property.

From the vSphere docs:

> If set true, memory resource reservation for this virtual machine will always be equal to the virtual machine's memory size; increases in memory size will be rejected when a corresponding reservation increase is not possible. This feature may only be enabled if it is currently possible to reserve all of the virtual machine's memory

File: src/vsphere_cpi/spec/integration/multi_cluster_overflow_spec.rb [hunk 2]
```diff
@@ -0,0 +1,123 @@
+require 'integration/spec_helper'
```
Comment (cunnie) [inline src/vsphere_cpi/spec/integration/multi_cluster_overflow_spec.rb:1]: I **love** this integration test.

**Review by cunnie — APPROVED** [review]
Nice job!

**Review by cunnie — COMMENTED** [review]
I'd amend the commit message slightly to show the use case (_why_ someone would use this), e.g.:

> Setting memory_reservation_locked_to_max: true might prove helpful when excessive VM migration is seen, evidenced by kernel messages such as "CPU#2 stuck for 23s!". This can happen when vSphere HA is enabled with admission control on clusters with constrained memory resources.

**Review by cunnie — APPROVED** [review]
LGTM

**Review by nouseforaname — CHANGES_REQUESTED** [review]
Other than the description of the 2nd test case, lgtm

File: src/vsphere_cpi/lib/cloud/vsphere/vm_type.rb (hunk 1 above)
Comment (rkoster) [reply src/vsphere_cpi/lib/cloud/vsphere/vm_type.rb:18]: Wat ahead of you 🙃 https://github.com/cloudfoundry/docs-bosh/pull/779

File: src/vsphere_cpi/spec/integration/multi_cluster_overflow_spec.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  context 'when the second cluster has enough reservable memory' do
+    before do
+      update_resource_pool_memory_reservation(cpi, @first_cluster_name, @first_rp, 100, false)
+      update_resource_pool_memory_reservation(cpi, @second_cluster_name, @second_rp, 1024, false)
+    end
+    after do
+      update_resource_pool_memory_reservation(cpi, @first_cluster_name, @first_rp, 1, true)
+      update_resource_pool_memory_reservation(cpi, @second_cluster_name, @second_rp, 1, true)
+    end
+    it 'should create a VM on the first cluster' do
```
[the remark this answers is not in the document]
Comment (rkoster) [reply src/vsphere_cpi/spec/integration/multi_cluster_overflow_spec.rb]: Fixed

**Review by nouseforaname — APPROVED** [review]
thanks!

File: src/vsphere_cpi/spec/integration/multi_cluster_overflow_spec.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (aramprice) [reply src/vsphere_cpi/spec/integration/multi_cluster_overflow_spec.rb]: Seems so

**Review by aramprice — APPROVED** [review]
Looks good to me. Seems like there's one spec description that should be updated at
> src/vsphere_cpi/spec/integration/multi_cluster_overflow_spec.rb:102

`first cluster` => `second cluster`

**Pull request merged**